### ISSUE TITLE
Translation in visual mode removes newlines

### ISSUE THREAD

**Issue by askfiy** [uga-rosa/translate.nvim#4]
For example I have the following string:

``` 
- 1. hello
- 2. world
```

When I selected 1 and 2 to translate or replace, it gets this:

```
- 1. result - 2. result
```

Is there any way to make it output as-is without removing the newlines?

**Comment by uga-rosa** [comment]
It is possible to separate them so that they can be configured. (hard coding now)
Basically, when translating multiple lines at once, they are not bullet points but one continuous sentence, and a normal sentence would have periods. I did it this way because it makes more sense to remove the line breaks.

I'm now implementing a function to select the comment under the cursor, but if Lua, for example, we need to delete the `--` on each line, so let's separate this function.

**Comment by askfiy** [comment]
Is it possible to add an option to let users choose whether they want or not to combine multiple lines of translations resulat into one line?

This helps when doing replaces.

People write comments in their native language, then translate the comments into English and publish them on github.

Manually re-tapping the Enter key to wrap lines can be cumbersome.

**Comment by uga-rosa** [comment]
It is difficult to handle `replace` because the number of characters changes before and after the translation. Now I manage to deal with it by putting it on one line. This plugin allows you to freely choose the range of translation. Do you have any good ideas?

**Comment by askfiy** [comment]
I don't know if this proposal is acceptable to you.

When using the relpace command in V mode, it is possible to go through each line in a loop, and perform a normal replace on each line, so that the result is not merged into one line.

This is the only possible solution I can think of until I come across a better solution.

Seems like vscode's comment translate solves this, I can ask the author of this plugin how it does it.

**Comment by askfiy** [comment]
Such a traversal is undoubtedly stupid, but it does solve the problem effectively.

It is up to the user to choose whether or not to enable this feature, as this will bring some performance penalty, although the number of lines that need to be translated or replaced is not many.

I think this performance penalty is slight, since users don't select too much content to translate at once.

But the benefit it brings is significant, we save time hitting enter.

**Comment by askfiy** [comment]
I have shared this plug-in on some famous platforms in China, and I believe that more people will use this plug-in.

**Comment by uga-rosa** [comment]
Thank you! There is still a high possibility of destructive changes coming in, so it would be helpful if you could pass that along as well.
I have a few ideas and will try to implement them tomorrow so I can test them out.

**Comment by askfiy** [comment]
Hopefully it's not just implemented under -output=replace, -outputu=split is also required.

Thanks again for bringing us such a handy plugin, it's definitely the most convenient plugin I've ever used on neovim.

**Comment by uga-rosa** [comment]
Changed `parse` to `parse_after` and added `parse_before.` Also added `mode` option to `output/replace` and `output/split.` Please try it out.

**Comment by askfiy** [comment]
> Changed `parse` to `parse_after` and added `parse_before.` Also added `mode` option to `output/replace` and `output/split.` Please try it out.

Changed parse to parse_after and added parse_before. Also added mode option to output/replace and output/split. Please try it out.

**Comment by askfiy** [comment]
ok i'll give it a try, thank you very much

**Comment by uga-rosa** [comment]
If you change `sep`, the option of `parse_before/concat,` to `\n`, you should be able to pass it to the translation command with newlines included. I can't be responsible for whether or not the newlines remain in the result.
And since we are not translating line by line, it is impossible in principle to insert line breaks at semantic breaks.

**Comment by uga-rosa** [comment]
I feel that the process of splitting the string of the translation result into an array should be put in parse_after. But I want the string as is for register only, and since floating and replacing require different processing, the number of arguments per command will increase, but it doesn't matter if it is mapped.

**Comment by askfiy** [comment]
Is it a problem with my configuration? I don't think \n behaves very well, is this normal?

![image](https://user-images.githubusercontent.com/81478335/157870209-8d7d4a19-e004-4f7d-ad0e-21ccd6010ad6.png)

**Comment by askfiy** [comment]
Also, there is a new bug where using the replace function throws an exception when nothing is selected.

```
Error executing vim.schedule lua callback: ...t/translate.nvim/lua/translate/preset/output/replace.lua:29: attempt to perform arithmetic on field 'row' (a nil value)
stack traceback:
        ...t/translate.nvim/lua/translate/preset/output/replace.lua:29: in function 'output'
        ...te/pack/packer/opt/translate.nvim/lua/translate/init.lua:86: in function 'cb'
        vim.lua:285: in function <vim.lua:285>
```

**Comment by uga-rosa** [comment]
Unfortunately, it is normal. Since the function to respect indentation has not yet been included, it must be adjusted to be the original width, including indentation. Try rectangle selection for now.

**Comment by askfiy** [comment]
- 1. hello.
- 2. world.

Is it possible to store all the results in a table with the row number as the key in after?

```
{
   1: "- 1. hello",
   2: "- 2. world"
}
```

We can take the whole message and translate it:

```
trans "{ 1: '-1. hello', 2: '-2. world'}"                                                                                                                                                                                ✔  
```

The final result:

```
{ 1：'-1。你好'，2：'-2。世界'}
```

Then you can look up the line number and replace it.

When replacing, you can find the first valid character of the line, and then replace. 

In this way, indentation and line breaks can be solved

**Comment by uga-rosa** [comment]
Please don't be absurd.
We are translating languages with different grammars, so it is impossible to deal with positions.

**Comment by uga-rosa** [comment]
If there is such a clear break in meaning, it should be translated separately.

**Comment by askfiy** [comment]
I'll ask the author of the vscode translation plugin to see how he does it.

**Comment by uga-rosa** [comment]
I am a Deepl API Pro user, and the development process was to handle this conveniently from neovim in the first place. So I don't intend to hit the Google translate API directly. I intend to handle the translation results without losing generality.

**Comment by askfiy** [comment]
I understand my dear, but their principles are the same. The google API also does not guarantee indentation and line breaks, which need to be handled manually.

**Comment by uga-rosa** [comment]
Rest assured that I intend to add the ability to take indentation into account. The recognition of the comment area is also taking much more shape. I plan to have it all finished by next week.

**Comment by askfiy** [comment]
You are awesome dear, you have saved all neovim users in Asia, Africa with bad English. I am proud of you

**Comment by askfiy** [comment]
Having said that, replace in normal mode cannot be used now.

How are you over there?

```
Error executing vim.schedule lua callback: ...t/translate.nvim/lua/translate/preset/output/replace.lua:29: attempt to perform arithmetic on field 'row' (a nil value)
stack traceback:
        ...t/translate.nvim/lua/translate/preset/output/replace.lua:29: in function 'output'
        ...te/pack/packer/opt/translate.nvim/lua/translate/init.lua:86: in function 'cb'
        vim.lua:285: in function <vim.lua:285>
```

**Comment by uga-rosa** [comment]
Hmmm, it was working on my end, so maybe I made a mistake when I cut the patch. I will fix it tomorrow.

**Comment by askfiy** [comment]
ok you are awesome

**Comment by uga-rosa** [comment]
Destructive change; parse_after and output specifications changed. All formatting and line splitting of the translation result is now done by parse_after.

**Comment by uga-rosa** [comment]
Example
```vim
:Translate ZH -parse_after=remove_newline,head -output=split
```

**Comment by uga-rosa** [comment]
I haven't fixed the command completion yet, so don't trust it and look at help. Please let me know if there is anything you don't understand or that seems wrong.

**Comment by askfiy** [comment]
![image](https://user-images.githubusercontent.com/81478335/158017711-4fff2d72-42a6-41b4-b988-764528ced6d5.png)

**Comment by uga-rosa** [comment]
Something is wrong.
Can you post that text so I can copy it so I can try it at hand?

**Comment by askfiy** [comment]
I don't know exactly how it does it, but it seems to work well and is easier for people to read.

![image](https://user-images.githubusercontent.com/81478335/158017953-e5286402-5f80-49be-9a42-d3d069c522ad.png)

**Comment by askfiy** [comment]
> Something is wrong. Can you post that text so I can copy it so I can try it at hand?

Is this an unsolvable problem?

The number of split lines of the translation result cannot always be consistent with the original number of lines.

In other words, the newline is in the wrong place.

How did you differentiate between newlines?

**Comment by uga-rosa** [comment]
I don't know how to align the line breaks with the original text. I don't know which character in the translation result corresponds to which character before translation. Because the entire selection is made into one string and thrown into the translator. So, the current implementation is only looking at the display width of the original text to isolate the translation result.

**Comment by uga-rosa** [comment]
If there are blank lines in between, it may be possible to separate them and throw them into the translator separately.
But I think it would be quicker to use a macro to repeat `vip<your translation mapping>`.

**Comment by askfiy** [comment]
I understand, do you think this feature is important?

If you find this feature necessary, I can explore how it does it.

As things stand, it seems that it can only be put aside for now.

If there are any new developments in the future, we will contact you again.

**Comment by uga-rosa** [comment]
I understand the desire to translate all at once, but the translation is done asynchronously, and I don't see a practical problem with that. However, it is not nice to have a character in a line that should have zero characters, so I will fix it.
I don't really like it when plug-ins do things "nice" on their own. I think the good thing about (neo)vim is that users can control many features themselves.

**Comment by askfiy** [comment]
Aren't you going to add it to [awesome-neovim](https://github.com/rockerBOO/awesome-neovim)?

**Comment by uga-rosa** [comment]
Let me an example of adding a grammatically meaningful line break when translating into English.
```lua
require("translate").setup({
    parse_after = {
        split_by_period = {
            cmd = function (text, _)
                local lines = {}
                local i = 0
                for line in string.gmatch(text, ".-[.?]") do
                    line = vim.trim(line)
                    table.insert(lines, line)
                end
                return lines
            end
        }
    },
})
```

And do this command
```vim
:'<,'>Translate EN -parse_after=remove_newline,split_by_period -output=split
```

This is result.
![image](https://user-images.githubusercontent.com/82267684/158019061-869cc615-f018-4068-896e-6ea6d2447c1b.png)

**Comment by uga-rosa** [comment]
> Aren't you going to add it to [awesome-neovim](https://github.com/rockerBOO/awesome-neovim)?

Only after the addition of features has settled in.

**Comment by askfiy** [comment]
> Let me an example of adding a grammatically meaningful line break when translating into English.
> 
> ```lua
> require("translate").setup({
>     parse_after = {
>         split_by_period = {
>             cmd = function (text, _)
>                 local lines = {}
>                 local i = 0
>                 for line in string.gmatch(text, ".-[.?]") do
>                     line = vim.trim(line)
>                     table.insert(lines, line)
>                 end
>                 return lines
>             end
>         }
>     },
> })
> ```
> 
> And do this command
> 
> ```viml
> :'<,'>Translate EN -parse_after=remove_newline,split_by_period -output=split
> ```
> 
> This is result. ![image](https://user-images.githubusercontent.com/82267684/158019061-869cc615-f018-4068-896e-6ea6d2447c1b.png)

这个看起来很棒，它是通过特定的字符结尾做到的吗？

**Comment by uga-rosa** [comment]
Yes, it's a function of `string.gmatch`, I post a [link](https://www.lua.org/manual/5.1/manual.html#pdf-string.gmatch) to lua's reference manual.

**Comment by uga-rosa** [comment]
Ah, the extra `local i = 0` remains. I deleted it.

**Comment by askfiy** [comment]
I noticed this.

Well, hopefully there will be a better way to solve it in the future, you are really patient, thank you very much.

**Comment by uga-rosa** [comment]
The ability for users to easily add features like this is the main feature of this plugin ;)

**Comment by uga-rosa** [comment]
I guess I'll close this issue for now. Please let me know if you have any further questions/feature requests.

**Comment by uga-rosa** [comment]
I added `parse_after/split_newline`.
If you set `\n` to option `sep` for `parse_before/concat` and use this instead of `parse_after/remove_newline`, you would get properly line breaks.
NOTE: Translation of passages where multiple lines form a single sentence would be unnatural.
